### ISSUE TITLE
Idea to reduce latency

### ISSUE THREAD

**Issue by vjbaker88** [dev-0x7C6/meta-retro#3]
Hi

Thanks for your work.
To reduce lantency (audio overhead) did you try to use RT kernel (ex: preempt rt) or Desktop low latency in kernel options ?
Did you try to replace alsa by jackd 2 server or tune pulseaudio paramaters ? 

PS: where is defined retro-core-image ? i didn't find .bb file like other images.

Regards

**Comment by dev-0x7C6** [comment]
Hi,

> To reduce lantency (audio overhead) did you try to use RT kernel (ex: preempt rt) or Desktop low latency in kernel options ?

Neither of them, but I'm on it. I'll try with `yocto-linux-rt` for intel machine and probably something for raspberry. 

> Did you try to replace alsa by jackd 2 server or tune pulseaudio paramaters ?

Only with stock pulseaudio configuration. I never tried jack I wonder can it coexists with pulseaudio without tons of configuration? I'll read on the matter.

Recently I was messing with retroarch configuration. I changed default `64ms` audio latency to `256ms` with 
 `RETROARCH_CONFIG_AUDIO_LATENCY` variable. This results with better performance for some emulators for lower end sbc's. Not sure if this was a good choice in general. 

> PS: where is defined retro-core-image ? i didn't find .bb file like other images.

Here: [retro-core-image.bbclass](https://github.com/dev-0x7C6/meta-retro/blob/master/classes/retro-core-image.bbclass) it's defined as template for `retro-image-full`,  `retro-image-standard` and  `retro-image-minimal`

What is your current sbc that you're working on?

Cheers

**Comment by vjbaker88** [comment]
I'm working on Raspberry pi 3 and 4. 
I'm trying to compile pi3 (32 and 64 bits) minimal image.
I'm blocked on compilation with LLVM 10 or spirv tools. Are you sure that Mesa - Gallium LLVM is not in conflict for raspberry ?

**Comment by dev-0x7C6** [comment]
> I'm working on Raspberry pi 3 and 4.
> I'm trying to compile pi3 (32 and 64 bits) minimal image.

I'm using [retro-build-environment](https://github.com/dev-0x7C6/retro-build-environment) for my rpi 3/4 builds, may come handy.

> I'm blocked on compilation with LLVM 10 or spirv tools. Are you sure that Mesa - Gallium LLVM is not in conflict for raspberry ?

You can share an error log (with verbose logs on) maybe I'll be able to help. I'm not 100% sure but I think that llvm gallium is not needed for raspberry so you can disable it to avoid dependencies.

Are you using my distribution configuration? ([retro.conf](https://github.com/dev-0x7C6/meta-retro/blob/master/conf/distro/retro.conf)) - if yes there is `RETRO_ADDITIONAL_MESA_GPUS` variable to specified what append additionally (take note that is only adding to mesa configuration and living it empty will not remove some defaults). You can override this from `local.conf`

In recent commit 0abee52b640eed85c17102ab07614056d4c11f18 I've added  `linux-raspberrypi-rt` kernel (5.4.72 with RT patches, and configuration that enabling preempt rt)

I can be set in `local.conf` like that: `PREFERRED_PROVIDER_virtual/kernel = "linux-raspberrypi-rt"`

**Comment by dev-0x7C6** [comment]
> In recent commit a4f1ae5 I've added linux-raspberrypi-rt kernel (5.4.72 with RT patches, and configuration that enabling preempt rt)
> I can be set in local.conf like that: PREFERRED_PROVIDER_virtual/kernel = "linux-raspberrypi-rt"

I'm moving kernels and other recipes specific to `raspberrypi` to separate [meta-raspberrypi-extras](https://github.com/dev-0x7C6/meta-raspberrypi-extras) layer.

**Comment by vjbaker88** [comment]
Hello

This is the error that i meet when i'm compiling spirv-tools.
I m using a rpi3-64, is there a way in retroarch.inc to choose kms and opengles ?
It seems that glslang is enabled and so spirv-tools is needed.

Regards

| [16/19] /home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/recipe-sysroot-native/usr/bin/aarch64-poky-linux/aarch64-poky-linux-g++  -DSPIRV_CHECK_CONTEXT -DSPIRV_COLOR_TERMINAL -DSPIRV_LINUX -DSPIRV_TIMER_ENABLED -I/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git -I/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/external/spirv-headers/include -I/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/include -I/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/test -I. -I/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/external/effcee -I/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/external/effcee/effcee/.. -I/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/external/re2 -isystem /home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/external/googletest/googletest/include -isystem /home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/external/googletest/googlemock/include -isystem /home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/external/googletest/googlemock -isystem /home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/external/googletest/googletest -mcpu=cortex-a53+crc+crypto -fstack-protector-strong  -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security  --sysroot=/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/recipe-sysroot  -O2 -pipe -g -feliminate-unused-debug-types -fmacro-prefix-map=/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0=/usr/src/debug/spirv-tools/git-r0                      -fdebug-prefix-map=/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0=/usr/src/debug/spirv-tools/git-r0                      -fdebug-prefix-map=/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/recipe-sysroot=                      -fdebug-prefix-map=/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/recipe-sysroot-native=  -fvisibility-inlines-hidden  -mcpu=cortex-a53+crc+crypto -fstack-protector-strong  -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security  --sysroot=/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/recipe-sysroot -g -fPIE   -Wall -Wextra -Wnon-virtual-dtor -Wno-missing-field-initializers -std=c++11 -fno-exceptions -fno-rtti -Wno-long-long -Wshadow -Wundef -Wconversion -Wno-sign-conversion -Wno-undef -Wno-shadow -std=gnu++11 -MD -MT test/opt/CMakeFiles/test_opt.dir/fold_test.cpp.o -MF test/opt/CMakeFiles/test_opt.dir/fold_test.cpp.o.d -o test/opt/CMakeFiles/test_opt.dir/fold_test.cpp.o -c /home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/test/opt/fold_test.cpp
| FAILED: test/opt/CMakeFiles/test_opt.dir/fold_test.cpp.o
| /home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/recipe-sysroot-native/usr/bin/aarch64-poky-linux/aarch64-poky-linux-g++  -DSPIRV_CHECK_CONTEXT -DSPIRV_COLOR_TERMINAL -DSPIRV_LINUX -DSPIRV_TIMER_ENABLED -I/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git -I/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/external/spirv-headers/include -I/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/include -I/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/test -I. -I/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/external/effcee -I/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/external/effcee/effcee/.. -I/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/external/re2 -isystem /home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/external/googletest/googletest/include -isystem /home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/external/googletest/googlemock/include -isystem /home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/external/googletest/googlemock -isystem /home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/external/googletest/googletest -mcpu=cortex-a53+crc+crypto -fstack-protector-strong  -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security  --sysroot=/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/recipe-sysroot  -O2 -pipe -g -feliminate-unused-debug-types -fmacro-prefix-map=/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0=/usr/src/debug/spirv-tools/git-r0                      -fdebug-prefix-map=/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0=/usr/src/debug/spirv-tools/git-r0                      -fdebug-prefix-map=/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/recipe-sysroot=                      -fdebug-prefix-map=/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/recipe-sysroot-native=  -fvisibility-inlines-hidden  -mcpu=cortex-a53+crc+crypto -fstack-protector-strong  -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security  --sysroot=/home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/recipe-sysroot -g -fPIE   -Wall -Wextra -Wnon-virtual-dtor -Wno-missing-field-initializers -std=c++11 -fno-exceptions -fno-rtti -Wno-long-long -Wshadow -Wundef -Wconversion -Wno-sign-conversion -Wno-undef -Wno-shadow -std=gnu++11 -MD -MT test/opt/CMakeFiles/test_opt.dir/fold_test.cpp.o -MF test/opt/CMakeFiles/test_opt.dir/fold_test.cpp.o.d -o test/opt/CMakeFiles/test_opt.dir/fold_test.cpp.o -c /home/gerald/Documents/poky/build/tmp/work/aarch64-poky-linux/spirv-tools/git-r0/git/test/opt/fold_test.cpp
| {standard input}: Assembler messages:
| {standard input}:2388687: Warning: end of file not at end of a line; newline inserted
| {standard input}:2388988: Error: symbol `.LVU5319' is already defined
| {standard input}: Error: open CFI at the end of file; missing .cfi_endproc directive
| aarch64-poky-linux-g++: fatal error: Killed signal terminated program cc1plus
| compilation terminated.

**Comment by vjbaker88** [comment]
On spirv-tools git, there is an update since 2 days to correct VUID label.
I'm checking.

**Comment by dev-0x7C6** [comment]
> This is the error that i meet when i'm compiling spirv-tools.

I'm using default version from `meta-oe`/`poky` which is `spirv-tools 2020.5` (for master branch) right now. Did you switched to git version because of compilation problem? Maybe it's good idea to revert to default one?

> is there a way in retroarch.inc to choose kms and opengles ?

Defaults are here: [retroarch-config.inc](https://github.com/dev-0x7C6/meta-retro/blob/master/recipes-retroarch/retroarch/retroarch-config.inc)

`kms` with `opengl (es)` is set by default. I recommend to stay with default configuration provided by recipe. I'm using `kms` back-end on couple of sbc and intel machines with this configuration.

> {standard input}:2388988: Error: symbol `.LVU5319' is already defined

Unfortunately I don't know how to deal with that one.

> On spirv-tools git, there is an update since 2 days to correct VUID label.

Give it a try, but I think it's better to stay with released/stable version.

**Comment by dev-0x7C6** [comment]
```
~/.../yocto/master/retro-build-environment ⑂master* $ MACHINE="raspberrypi3-64" bitbake retro-image-full
Loading cache: 100% |########################################################################################################################################################################################################################| Time: 0:00:00
Loaded 4360 entries from dependency cache.
Parsing recipes: 100% |######################################################################################################################################################################################################################| Time: 0:00:03
Parsing of 2964 .bb files complete (2803 cached, 161 parsed). 4470 targets, 363 skipped, 0 masked, 0 errors.
NOTE: Resolving any missing task queue dependencies

Build Configuration:
BB_VERSION           = "1.49.0"
BUILD_SYS            = "x86_64-linux"
NATIVELSBSTRING      = "universal"
TARGET_SYS           = "aarch64-poky-linux"
MACHINE              = "raspberrypi3-64"
DISTRO               = "retro"
DISTRO_VERSION       = "3.2+snapshot-20201107"
TUNE_FEATURES        = "aarch64 armv8a crc cortexa53"
TARGET_FPU           = ""
meta-dev1990         = "master:a1d6c798d027b89c24537d4b43a835af4bb39875"
meta-intel           = "HEAD:3cf8482e432e5daf0b75394152e2c229b3a08d36"
meta-kodi            = "master-next:1bb0e6aa6a7008aa16a27e86e66fbce59b269d7d"
meta-mesa            = "master:7470c62fa9d9f537c734ce4f2dec9f249fae8746"
meta-meson           = "master:2cb8fcf6679fc3c1d0f07279413a4f7a20203534"
meta-multimedia      
meta-networking      
meta-oe              
meta-python          = "HEAD:1204ed0a0efa600d07cea7dd78036f051e44f107"
meta-python2         = "HEAD:c43c29e57f16af4e77441b201855321fbd546661"
meta-qt5             = "HEAD:6557244eca22a8722cf0fde2f3cc62f45cb72651"
meta-raspberrypi     = "HEAD:15739db1ed469babe6cd162f0c100cae50919dcd"
meta-raspberrypi-extras = "HEAD:ee0c01f9169360736a4ac061739cf914bf9f79e0"
meta-rauc            = "HEAD:8e049699f05483b694e6c5a3e260cd9e0fe58d2e"
meta-rauc-raspberrypi = "autogrow-fs:8273fad07c49d276e2963976f5911f84a8e4de32"
meta-retro           = "master:28bed52815d695dd40ccf9e5771b566702782c97"
meta-sunxi           = "HEAD:370007f1c8694b7fd1847287ad66d9916581540c"
meta                 
meta-poky            
meta-yocto-bsp       = "master:1ad28364512be087f35b7cac6f6289027c070767"

Initialising tasks: 100% |###################################################################################################################################################################################################################| Time: 0:00:09
Sstate summary: Wanted 745 Found 744 Missed 1 Current 2540 (99% match, 99% complete)
NOTE: Executing Tasks
NOTE: Tasks Summary: Attempted 8988 tasks of which 8966 didn't need to be rerun and all succeeded.
```

I was able to build `retro-image-full` for `raspberrypi3-64` without any errors.

**Comment by vjbaker88** [comment]
Thanks for the research. I'm using dunfell branch which seems to be outdated since 1 month.

Layer openembedded-layer is not compatible with the core layer which only supports these series: dunfell (layer is compatible with hardknott gatesgarth)
